front/lib/api/assistant/reaction.ts

```diff
@@ -0,0 +1,82 @@
+import { Authenticator } from "@app/lib/auth";
+import { ModelId } from "@app/lib/databases";
+import { Message, MessageReaction } from "@app/lib/models";
+import { MessageReactionType } from "@app/types/assistant/conversation";
+
+export async function addOrRemoveMessageReaction(
+  auth: Authenticator,
+  {
+    messageId,
+    userId,
+    userContextUsername,
+    userContextFullName,
+    reaction,
+  }: {
+    messageId: string;
+    userId: ModelId;
+    userContextUsername: string;
+    userContextFullName: string | null;
+    reaction: string;
+  }
+): Promise<boolean | null> {
+  const owner = auth.workspace();
+  if (!owner) {
+    throw new Error("Unexpected `auth` without `workspace`.");
+  }
+
+  const message = await Message.findOne({
+    where: {
+      sId: messageId,
+    },
+  });
+
+  if (!message) {
+    return null;
+  }
+
+  const existingReaction = await MessageReaction.findOne({
+    where: {
+      messageId: message.id,
+      userId,
+      reaction,
+    },
+  });
+  if (existingReaction) {
+    await existingReaction.destroy();
```

Comment:
Looks like you implemented the click an existing reaction to remove it all the way down the server side.

I think we should separate the two actions even if I agree this form is more compact. It's also non idempotent which is not great.

Reply:
That means the React component is responsive for calling POST or DELETE after checking the current user has reacted on a given emoji, making it more complex. I thought this way was better for when we want to introduce reactjis from Slack users. But I have no problem changing it!

Reply:
Yes. The big upside is that it's idempotent and can be retried. If you get an error with your current endpoint you don't quite know what to do next